examples/all-clusters-app/linux/diagnostic-logs-provider-delegate-impl.cpp

```diff
@@ -0,0 +1,155 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "include/diagnostic-logs-provider-delegate-impl.h"
+#include <app/clusters/diagnostic-logs-server/diagnostic-logs-provider-delegate.h>
+#include <app/clusters/diagnostic-logs-server/diagnostic-logs-server.h>
+
+using namespace chip;
+using namespace chip::app::Clusters::DiagnosticLogs;
+
+LogProvider LogProvider::sInstance;
+
+LogSessionHandle LogProvider::sLogSessionHandle;
+
+// TODO: Fix #30477 - Add support for multiple log collection sessions
+
+LogSessionHandle LogProvider::StartLogCollection(IntentEnum logType)
+{
+
+    // We can handle only one log collection session. Return kInvalidLogSessionHandle
+    // if we are in the middle of another log collection session.
+    if (mIsInALogCollectionSession)
+    {
+        return kInvalidLogSessionHandle;
+    }
+
+    mIsInALogCollectionSession  = true;
+    mTotalNumberOfBytesConsumed = 0;
+
+    // Open the log file for reading.
+    Optional<std::string> filePath = GetLogFilePath(logType);
+    if (filePath.HasValue())
+    {
+        mFileStream.open(filePath.Value().c_str(), std::ios_base::binary | std::ios_base::in);
+        if (!mFileStream.good())
+        {
+            ChipLogError(BDX, "Failed to open the log file");
+            return kInvalidLogSessionHandle;
+        }
+        sLogSessionHandle++;
+
+        // If the session handle rolls over to UINT16_MAX which is invalid, reset to 0.
+        if (sLogSessionHandle == kInvalidLogSessionHandle)
+        {
+            sLogSessionHandle = 0;
+        }
+        mLogSessionHandle = sLogSessionHandle;
+    }
+    else
+    {
+        mLogSessionHandle = kInvalidLogSessionHandle;
+    }
+    return mLogSessionHandle;
+}
+
+CHIP_ERROR LogProvider::GetNextChunk(LogSessionHandle logSessionHandle, MutableByteSpan & outBuffer, bool & outIsEOF)
+{
+    if (!mIsInALogCollectionSession)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    if (!mFileStream.is_open())
+    {
+        ChipLogError(BDX, "File is not open");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    mFileStream.seekg(mFileStream.end);
+    long long fileSize = mFileStream.tellg();
+
+    long long remainingBytesToBeRead = fileSize - static_cast<long long>(mTotalNumberOfBytesConsumed);
+
+    if ((remainingBytesToBeRead >= kMaxLogContentSize && outBuffer.size() < kMaxLogContentSize) ||
+        (remainingBytesToBeRead < kMaxLogContentSize && static_cast<long long>(outBuffer.size()) < remainingBytesToBeRead))
```

Comment:
I think this check should move to the calling code. The intention was to check if remaining bytes left are greater than BDX max block size (i used the incorrect constant here), then the buffer should have space for max block size. otherwise it should fit remaining bytes. I agree this is the incorrect spot to check. let me rethink and see if its even needed.